File: frontend/src/lib/core/api/simulation.ts
import {
	SimulationService,
	SimulationStatus,
	SimulationType,
	type ShadowPolygon as ProtoShadowPolygon,
	type Simulation as ProtoSimulation,
	type SunPosition as ProtoSunPosition
} from '$lib/gen/simulation/v1/simulation_pb.js';

import { createApiClient, isoToTimestamp, timestampToIso } from './connect';

export interface Simulation {
	id: string;
	project_id: string;
	layout_id: string;
	name: string;
	simulation_type: string;
	status: string;
	params: SimulationParams;
	result: SimulationResult | null;
	created_at: string;
	completed_at: string | null;
}

export interface SimulationParams {
	start_time: string;
	end_time: string;
	time_step_minutes: number;
	latitude: number;
	longitude: number;
	include_terrain_shading: boolean;
	include_panel_shading: boolean;
}

export interface SimulationResult {
	total_irradiance_kwh_m2: number;
	annual_yield_kwh: number;
	performance_ratio: number;
	shading_loss_percent: number;
	result_file_path: string;
}

export interface SunPosition {
	azimuth: number;
	elevation: number;
	zenith: number;
	hour_angle: number;
}

export interface ShadowPolygon {
	source_panel_id: string;
	shadow_geojson: string;
	shadow_intensity: number;
}

const client = createApiClient(SimulationService);

function simulationTypeFromProto(simulationType: SimulationType): string {
	switch (simulationType) {
		case SimulationType.SHADOW:
			return 'shadow';
		case SimulationType.IRRADIANCE:
			return 'irradiance';
		case SimulationType.ANNUAL_YIELD:
			return 'yield';
		default:
			return 'irradiance';
	}
}

function simulationTypeToProto(simulationType: string): SimulationType {
	switch (simulationType) {
		case 'shadow':
			return SimulationType.SHADOW;
		case 'yield':
			return SimulationType.ANNUAL_YIELD;
		case 'irradiance':
		default:
			return SimulationType.IRRADIANCE;
	}
}

function simulationStatusFromProto(status: SimulationStatus): string {
	switch (status) {
		case SimulationStatus.PENDING:
			return 'pending';
		case SimulationStatus.RUNNING:
			return 'running';
		case SimulationStatus.COMPLETED:
			return 'completed';
		case SimulationStatus.FAILED:
			return 'failed';
		default:
			return 'pending';
	}
}

function mapSimulation(simulation?: ProtoSimulation): Simulation {
	if (!simulation) {
		throw new Error('Simulation response was empty');
	}

	return {
		id: simulation.id,
		project_id: simulation.projectId,
		layout_id: simulation.layoutId,
		name: simulation.name,
		simulation_type: simulationTypeFromProto(simulation.simulationType),
		status: simulationStatusFromProto(simulation.status),
		params: {
			start_time: timestampToIso(simulation.params?.startTime),
			end_time: timestampToIso(simulation.params?.endTime),
			time_step_minutes: simulation.params?.timeStepMinutes ?? 0,
			latitude: simulation.params?.latitude ?? 0,
			longitude: simulation.params?.longitude ?? 0,
			include_terrain_shading: simulation.params?.includeTerrainShading ?? false,
			include_panel_shading: simulation.params?.includePanelShading ?? false
		},
		result: simulation.result
			? {
					total_irradiance_kwh_m2: simulation.result.totalIrradianceKwhM2,
					annual_yield_kwh: simulation.result.annualYieldKwh,
					performance_ratio: simulation.result.performanceRatio,
					shading_loss_percent: simulation.result.shadingLossPercent,
					result_file_path: simulation.result.resultFilePath
				}
			: null,
		created_at: timestampToIso(simulation.createdAt),
		completed_at: simulation.completedAt ? timestampToIso(simulation.completedAt) : null
	};
}

function mapSunPosition(position?: ProtoSunPosition): SunPosition {
	if (!position) {
		throw new Error('Sun position response was empty');
	}

	return {
		azimuth: position.azimuth,
		elevation: position.elevation,
		zenith: position.zenith,
		hour_angle: position.hourAngle
	};
}

function mapShadow(shadow: ProtoShadowPolygon): ShadowPolygon {
	return {
		source_panel_id: shadow.sourcePanelId,
		shadow_geojson: shadow.shadowGeojson,
		shadow_intensity: shadow.shadowIntensity
	};
}

export const simulationApi = {
	create: async (data: {
		project_id: string;
		layout_id: string;
		name: string;
		simulation_type: string;
		params: SimulationParams;
	}) => {
		const response = await client.createSimulation({
			projectId: data.project_id,
			layoutId: data.layout_id,
			name: data.name,
			simulationType: simulationTypeToProto(data.simulation_type),
			params: {
				startTime: isoToTimestamp(data.params.start_time),
				endTime: isoToTimestamp(data.params.end_time),
				timeStepMinutes: data.params.time_step_minutes,
				latitude: data.params.latitude,
				longitude: data.params.longitude,
				includeTerrainShading: data.params.include_terrain_shading,
				includePanelShading: data.params.include_panel_shading
			}
		});

		return { simulation: mapSimulation(response.simulation) };
	},

	get: async (id: string) => {
		const response = await client.getSimulation({ id });
		return { simulation: mapSimulation(response.simulation) };
	},

	list: async (projectId: string) => {
		const response = await client.listSimulations({ projectId });
		return { simulations: response.simulations.map(mapSimulation) };
	},

	run: async (id: string) => {
		const response = await client.runSimulation({ simulationId: id });
		return { simulation: mapSimulation(response.simulation) };
	},

	getSunPosition: async (lat: number, lon: number, timestamp: string) => {
		const response = await client.getSunPosition({
			latitude: lat,
			longitude: lon,
			timestamp: isoToTimestamp(timestamp)
		});

		return { position: mapSunPosition(response.position) };
	},

	getShadowMap: async (layoutId: string, timestamp: string, lat: number, lon: number) => {
		const response = await client.getShadowMap({
			layoutId,
			timestamp: isoToTimestamp(timestamp),
			latitude: lat,
			longitude: lon
		});

		return {
			shadows: response.shadows.map(mapShadow),
			sun_position: response.sunPosition ? mapSunPosition(response.sunPosition) : null
		};
	},

	delete: async (id: string) => {
		await client.deleteSimulation({ id });
		return {};
	}
};

File: mobile/lib/core/models/simulation.dart
import 'package:equatable/equatable.dart';

class Simulation extends Equatable {
  final String id;
  final String projectId;
  final String layoutId;
  final String name;
  final String simulationType;
  final String status;
  final SimulationParams params;
  final SimulationResult? result;
  final DateTime createdAt;
  final DateTime? completedAt;

  const Simulation({
    required this.id,
    required this.projectId,
    required this.layoutId,
    required this.name,
    required this.simulationType,
    required this.status,
    required this.params,
    this.result,
    required this.createdAt,
    this.completedAt,
  });

  factory Simulation.fromJson(Map<String, dynamic> json) {
    return Simulation(
      id: json['id'] as String,
      projectId: json['project_id'] as String,
      layoutId: json['layout_id'] as String,
      name: json['name'] as String,
      simulationType: json['simulation_type'] as String,
      status: json['status'] as String,
      params:
          SimulationParams.fromJson(json['params'] as Map<String, dynamic>),
      result: json['result'] != null
          ? SimulationResult.fromJson(json['result'] as Map<String, dynamic>)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
    );
  }

  bool get isCompleted => status == 'COMPLETED';
  bool get isRunning => status == 'RUNNING';
  bool get isPending => status == 'PENDING';
  bool get isFailed => status == 'FAILED';

  @override
  List<Object?> get props => [id, status, simulationType];
}

class SimulationParams extends Equatable {
  final String startTime;
  final String endTime;
  final int timeStepMinutes;
  final double latitude;
  final double longitude;
  final bool includeTerrainShading;
  final bool includePanelShading;

  const SimulationParams({
    required this.startTime,
    required this.endTime,
    this.timeStepMinutes = 60,
    required this.latitude,
    required this.longitude,
    this.includeTerrainShading = true,
    this.includePanelShading = true,
  });

  factory SimulationParams.fromJson(Map<String, dynamic> json) {
    return SimulationParams(
      startTime: json['start_time'] as String,
      endTime: json['end_time'] as String,
      timeStepMinutes: json['time_step_minutes'] as int? ?? 60,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      includeTerrainShading: json['include_terrain_shading'] as bool? ?? true,
      includePanelShading: json['include_panel_shading'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        'start_time': startTime,
        'end_time': endTime,
        'time_step_minutes': timeStepMinutes,
        'latitude': latitude,
        'longitude': longitude,
        'include_terrain_shading': includeTerrainShading,
        'include_panel_shading': includePanelShading,
      };

  @override
  List<Object?> get props => [startTime, endTime, latitude, longitude];
}

class SimulationResult extends Equatable {
  final double totalIrradianceKwhM2;
  final double annualYieldKwh;
  final double performanceRatio;
  final double shadingLossPercent;
  final String? resultFilePath;

  const SimulationResult({
    required this.totalIrradianceKwhM2,
    required this.annualYieldKwh,
    required this.performanceRatio,
    required this.shadingLossPercent,
    this.resultFilePath,
  });

  factory SimulationResult.fromJson(Map<String, dynamic> json) {
    return SimulationResult(
      totalIrradianceKwhM2:
          (json['total_irradiance_kwh_m2'] as num).toDouble(),
      annualYieldKwh: (json['annual_yield_kwh'] as num).toDouble(),
      performanceRatio: (json['performance_ratio'] as num).toDouble(),
      shadingLossPercent: (json['shading_loss_percent'] as num).toDouble(),
      resultFilePath: json['result_file_path'] as String?,
    );
  }

  @override
  List<Object?> get props =>
      [totalIrradianceKwhM2, annualYieldKwh, performanceRatio];
}

class SunPosition extends Equatable {
  final double azimuth;
  final double elevation;
  final double zenith;
  final double hourAngle;

  const SunPosition({
    required this.azimuth,
    required this.elevation,
    required this.zenith,
    required this.hourAngle,
  });

  factory SunPosition.fromJson(Map<String, dynamic> json) {
    return SunPosition(
      azimuth: (json['azimuth'] as num).toDouble(),
      elevation: (json['elevation'] as num).toDouble(),
      zenith: (json['zenith'] as num).toDouble(),
      hourAngle: (json['hour_angle'] as num).toDouble(),
    );
  }

  @override
  List<Object?> get props => [azimuth, elevation];
}
